# Bài tập 6

Bùi Chí Dũng - 1712364

## Câu 1 (1 điểm)

Nếu sử dụng tập hợp con H' của hypothesis H thì sẽ tăng bias, tức là tăng deterministic noise. Vì vậy em chọn [b] In general, deterministic noise will increase.

## Câu 2 (1 điểm)

In [1]:
import numpy as np
train_data = np.loadtxt("in.dta")
train_X, train_Y = train_data[:,:2], train_data[:,2]

test_data = np.loadtxt("out.dta")
test_X, test_Y = test_data[:,:2], test_data[:,2]

In [2]:
def phi(X):
    Z = np.zeros((X.shape[0],  8))
    Z[:,0] = 1
    Z[:,1] = X[:,0]
    Z[:,2] = X[:,1]
    Z[:,3] = X[:,0]**2
    Z[:,4] = X[:,1]**2
    Z[:,5] = X[:,0] * X[:,1]
    Z[:,6] = np.abs(X[:,0] - X[:,1])
    Z[:,7] = np.abs(X[:,0] + X[:,1])
    return Z

# Transform the data
train_Z = phi(train_X)

# Fit the data
Z = train_Z

#w_reg(slide 11 - Regularization) với 𝜆 = 0
w_t = np.linalg.solve(Z.T @ Z, Z.T @ train_Y)

def predict(X):
    Z = phi(X)
    return np.sign(Z @ w_t)

In [3]:
#Compute E_in
E_in = np.mean(predict(train_X) != train_Y)
print("E_in = ", E_in)

#Compute E_out
E_out = np.mean(predict(test_X) != test_Y)
print("E_out = ", E_out)

E_in =  0.02857142857142857
E_out =  0.084


Dựa vào kết quả trên, em chọn [a] 0.03, 0.08

## Câu 3 (1 điểm)

In [4]:
k = -3
# Data transf
train_Z = phi(train_X)
# Compute w
Z = train_Z
w_t = np.linalg.solve(Z.T @ Z + 10**k*np.identity(Z.shape[1]), Z.T @ train_Y)

def predict(X):
    Z = phi(X)
    return np.sign(Z @ w_t)

E_in = np.mean(predict(train_X) != train_Y)
print("E_in =", E_in)
    
E_out = np.mean(predict(test_X) != test_Y)
print("E_out =", E_out)

E_in = 0.02857142857142857
E_out = 0.08


Từ kết quả trên, ta thấy E_in không thay đổi so với ở câu 2, E_out có cải thiện hơn khi giảm từ 0.084 -> 0.08. Chọn [d] 0.03, 0.08

## Câu 4 (1 điểm)

In [5]:
k = 3
# Data transf
train_Z = phi(train_X)
# Compute w
Z = train_Z
w_t = np.linalg.solve(Z.T @ Z + 10**k*np.identity(Z.shape[1]), Z.T @ train_Y)

def predict(X):
    Z = phi(X)
    return np.sign(Z @ w_t)
E_in = np.mean(predict(train_X) != train_Y)
print("E_in =", E_in)
    
E_out = np.mean(predict(test_X) != test_Y)
print("E_out =", E_out)

E_in = 0.37142857142857144
E_out = 0.436


Với k = 3, ta thấy kết quả gần với câu [e] 0.4, 0.4

## Câu 5 (1 điểm)

In [6]:
# Data transf
train_Z = phi(train_X)

for k in range(-2,2):
    # Compute w
    Z = train_Z
    w_t = np.linalg.solve(Z.T @ Z + 10**k*np.identity(Z.shape[1]), Z.T @ train_Y)

    def predict(X):
        Z = phi(X)
        return np.sign(Z @ w_t)
    print("Với k =", k)
    E_in = np.mean(predict(train_X) != train_Y)
    print("E_in =", E_in)
    E_out = np.mean(predict(test_X) != test_Y)
    print("E_out =", E_out)

Với k = -2
E_in = 0.02857142857142857
E_out = 0.084
Với k = -1
E_in = 0.02857142857142857
E_out = 0.056
Với k = 0
E_in = 0.0
E_out = 0.092
Với k = 1
E_in = 0.05714285714285714
E_out = 0.124


Ta thấy với k = -1, thu được E_out nhỏ nhất ~= 0.056. Chọn [d] -1

## Câu 6 (1 điểm)

Ta thấy với k = -1, thu được E_out nhỏ nhất ~= 0.056. Chọn [b] 0.06

## Câu 7 (1 điểm)

Ta có H(10, 0, 3) = H2, bởi vì w = 0 với mọi Q >= 3, tức là từ L3, L4, ...,L(Q).

w có giá trị với L0, L1 và L2. 

Mặt khác ta thấy H(10, 0, 3) là tập con của H(10,0,4), vì vậy H(10,0,3) ∩ H(10,0,4) = H(10,0,3) = H2.
Chọn [c] H(10, 0, 3) ∩ H(10, 0, 4) = H2

## Câu 8 (1 điểm)

- Forward

layer1_1 = 3 * 6 = 18


layer2_1 = 1 * 4 = 4


operations1 = layer1_1 + layer2_1 = 22

- Backward, tính delta_l trong hidden layer + output layer

layer2_3 = 0 # delta_l output node là 0


layer1_3 = 3 # bias neuron không có delta_l


operations3 = layer2_3 + layer1_3 = 3

- Vòng lặp thực hiện cả backward hay forward: Tính de/dw(i,j)(l) = x(l-1)*delta_l(l) (operations3)

layer2_2 = 4


layer1_2 = 3 * 6


operations2 = layer2_2 + layer1_2 = 22

- Total: operations1 + operations2 + operations3 = 22 + 22 + 3 = 47

Chọn [d] 45

## Câu 9 (1 điểm)

Em chọn câu [a] 46

## Câu 10 (1 điểm)

Em chọn câu [e] 510